### PR TITLE
Enhance codecept command

### PR DESCRIPTION
**Pull request by shadyvb**
- Adds iteration over all suites and execute them one by one
- Detects use of WPDb / WPWebDriver and preps db/browser
- Adds `codecept bootstrap` command to copy template suites to a new directory

### CODE REVIEW

File: inc/command/class-command.php [hunk 1]
```diff
@@ -487,6 +518,58 @@ protected function run_command( InputInterface $input, OutputInterface $output,
...
+					)
+				);
+			}
+		}
+
+		$base_path = getcwd(); // Do we have another way to detect this ?
+		$tests_folder = $base_path . '/' . $tests_folder;
+
+		if ( ! file_exists( $tests_folder ) ) {
+			mkdir( $tests_folder );
```
Comment (roborourke) [inline inc/command/class-command.php]: Might need to specify the permissions and recursive arguments too, path might be nested:
```suggestion
			mkdir( $tests_folder, 0755, true );
```

File: inc/command/class-command.php [hunk 2]
```diff
@@ -487,6 +518,58 @@ protected function run_command( InputInterface $input, OutputInterface $output,
...
+		} else {
+			foreach ( $suites as $suite ) {
+				if ( file_exists( $tests_folder . '/' . $suite ) ) {
```
Comment (roborourke) [inline inc/command/class-command.php:515]: Because this part of the command is running on the host OS might need to use `DIRECTORY_SEPARATOR` instead, similar in a few other places too. Not sure if PHP just handles it or not on Windows.

File: inc/command/class-command.php [hunk 3]
```diff
@@ -663,6 +750,69 @@ protected function stop_browser_container( InputInterface $input, OutputInterfac
...
+	 * @param InputInterface $input
+	 *
+	 * @return string
+	 */
+	protected function get_test_suite_argument( InputInterface $input ) : string {
+		$arguments = $input->getArguments();
+		$first_argument = $arguments['options'][1] ?? '';
+
+		// Test suite should be the first argumet so we can check for it, otherwise, assume no suite is selected.
+		if ( 0 === strpos( $first_argument, '-' ) ) {
```
Comment (roborourke) [inline inc/command/class-command.php]: Not sure I follow why there'd definitely be a dash here. Shouldn't this method check if we've called `codecept run` too?

What I ran `composer dev-tools codecept g:cest acceptance SomeName` for example.

File: inc/command/class-command.php [hunk 4]
```diff
@@ -213,11 +214,18 @@ protected function phpunit( InputInterface $input, OutputInterface $output ) {
 	 */
 	protected function codecept( InputInterface $input, OutputInterface $output ) {
 		$options = $input->getArgument( 'options' );
-		$tests_folder = $input->getOption( 'path' );
+		// Codeception command runs within `vendor` directory, so paths are relative to that.
+		$tests_folder = '../' . $input->getOption( 'path' );
```
Comment (roborourke) [inline inc/command/class-command.php]: Probably worth sanitising this value to strip any trailing slashes or directory separators.

File: inc/command/class-command.php [hunk 5]
```diff
@@ -213,11 +214,18 @@ protected function phpunit( InputInterface $input, OutputInterface $output ) {
...
 		$output_folder = $input->getOption( 'output' );
-		$run_headless_browser = $input->getOption( 'browser' );
+		$run_headless_browser = $input->getOption( 'browser' ) ?: 'chrome';
```
Comment (roborourke) [inline inc/command/class-command.php]: Why not leave this as the default value for the option now that starting the browser is dependent on whether the suite uses the web driver module?

File: inc/command/class-command.php [hunk 6]
```diff
@@ -213,11 +214,18 @@ protected function phpunit( InputInterface $input, OutputInterface $output ) {
...
 		$use_chassis = $input->getOption( 'chassis' );
 		$project_subdomain = $this->get_project_subdomain();
+		$subsubcommand = $input->getArguments()['options'][0] ?? '';
+		$test_suite = $this->get_test_suite_argument( $input );
+
+		if ( $subsubcommand === 'bootstrap' ) {
```
Comment (roborourke) [inline inc/command/class-command.php]: Do we need to consider anything in relation to codecept's built in bootstrap command? This basically does the same thing but more tailored to our use case so not a major issue, just wanting to make a note.

File: inc/command/class-command.php [hunk 7]
```diff
@@ -213,11 +214,18 @@ protected function phpunit( InputInterface $input, OutputInterface $output ) {
...
 		$options = $input->getArgument( 'options' );
-		$tests_folder = $input->getOption( 'path' );
+		// Codeception command runs within `vendor` directory, so paths are relative to that.
+		$tests_folder = '../' . $input->getOption( 'path' );
 		$output_folder = $input->getOption( 'output' );
-		$run_headless_browser = $input->getOption( 'browser' );
+		$run_headless_browser = $input->getOption( 'browser' ) ?: 'chrome';
 		$use_chassis = $input->getOption( 'chassis' );
 		$project_subdomain = $this->get_project_subdomain();
+		$subsubcommand = $input->getArguments()['options'][0] ?? '';
```
Comment (roborourke) [inline inc/command/class-command.php]: For the most part this'll be fine but people could throw all sorts at it like `composer dev-tools codecept -p some/path bootstrap`. We can't always control the order options come in so if we're only looking for subsubcommand to match bootstrap this can probably just be an `in_array()` check.

File: inc/command/class-command.php [hunk 8]
```diff
@@ -487,6 +518,58 @@ protected function run_command( InputInterface $input, OutputInterface $output,
...
+	 * @param InputInterface $input
+	 * @param OutputInterface $output
+	 *
+	 * @return void
+	 */
+	protected function bootstrap_codecept( string $tests_folder, InputInterface $input, OutputInterface $output ) {
+		$default_suites = $this->get_test_suites( 'vendor/altis/dev-tools/tests' );
+
+		$suites = $input->getArguments()['options'][1] ?? '';
+		if ( 0 === strpos( $suites, '-' ) ) {
```
Comment (roborourke) [inline inc/command/class-command.php]: Still not sure where the dash would be coming from here or how reliable it is.

**Review by roborourke — COMMENTED** [review]
Liking it - I do want to be certain it's not too aggressive around the `get_test_suite_argument()`  part, all the other codecept commands should just work as intended ideally and it looks like some might unintentionally be handled incorrectly.

File: inc/command/class-command.php (hunk 2 above)
Comment (shadyvb) [reply inc/command/class-command.php:515]: PHP filesystem functions will always convert slashes to the appropriate directory separator automatically, so we don't really need to use `DIRECTORY_SEPARATOR` while compiling paths, maybe only for deconstructing paths coming from a system function, eg: exploding a path via the directory separator.

File: inc/command/class-command.php (hunk 6 above)
Comment (shadyvb) [reply inc/command/class-command.php]: Nothing that comes to mind, we can always revisit if something needs to be added.

File: inc/command/class-command.php (hunk 7 above)
Comment (shadyvb) [reply inc/command/class-command.php]: `InputInterface` does a good job clearing parameter arguments, so even in that case, `bootstrap` will still be returned as the first array element. An `in_array` call can match directories as well.

However, `hasParameterOption` seems to be cleaner, so will go with that.

File: inc/command/class-command.php (hunk 3 above)
Comment (shadyvb) [reply inc/command/class-command.php]: The goal here was to exclude any parameter argument, eg: `-p`, `--path`, etc.

I've done some more tests with a few variations, all seems to be returning the suite name without even checking if it's a parameter or option argument.

> composer dev-tools codecept g:cest acceptance SomeName
composer dev-tools codecept run acceptance
composer dev-tools codecept run -p some/path acceptance
composer dev-tools codecept -p some/path run acceptance

However, it'll fail if the suite name is passed after the end of options signal `--`, which I think is an edge case.

So I'll remove that condition.

Comment (shadyvb) [reply inc/command/class-command.php]: Ah, now I remember, it was these formats that might have failed, because the arguments after the end of options signal aren't parsed, unlike anything before it:

eg: this will fail:
> composer dev-tools codecept run -- -p test acceptance

this will work:
> composer dev-tools codecept run -p test acceptance

I'll try to find a way to mitigate that.

File: inc/command/class-command.php (hunk 2 above)
Comment (roborourke) [reply inc/command/class-command.php:515]: I thought it did something like that but couldn't remember for sure. Nice 👍

File: inc/command/class-command.php (hunk 7 above)
Comment (roborourke) [reply inc/command/class-command.php]: Ahh good to know, was thinking of it like the raw `$argv` data.

File: inc/command/class-command.php (hunk 3 above)
Comment (shadyvb) [reply inc/command/class-command.php]: I don't think we need to support `composer dev-tools codecept run -- -p test acceptance`, doesn't seem like it is a standard approach anyways.

File: inc/command/class-command.php (hunk 8 above)
Comment (shadyvb) [reply inc/command/class-command.php]: The dash represents a parameter argument, and the only case we're handling here is something that's passed beyond the end of options signal, eg: `codecept bootstrap -- -x something`.
`codecept bootstrap -- -x something suite1,suite2` seems like quite an edge case to expect, also non-standard approach.

File: inc/command/class-command.php [hunk 9]
```diff
@@ -41,14 +42,12 @@ protected function configure() {
...
-To run Codeception integration tests:
-    codecept [--chassis] -p <path> -b <browser> -o <output-folder> [--] [options]
-                                use `--` to separate arguments you want to
-                                pass to Codeception. Use the --chassis option
-                                if you are running Local Chassis. Use -p path
-                                to specify custom tests folder. Use --browser/-b
-                                to run a headless browser container for acceptance tests,
-                                choose 'chrome', or 'firefox' as needed.
+To run Codeception commands:
+    codecept <subcommand> -p <path> -b <browser> -o <output-folder> <suite-name> [--] [options]
```
Comment (roborourke) [inline inc/command/class-command.php]: Optional parameters should be wrapped in square brackets generally, which is pretty much everything after codecept I think:
```suggestion
    codecept [<subcommand>] [-p <path>] [-b <browser>] [-o <output-folder>] [<suite-name>] [--] [options]
```
Is suite name always required? Codeception has a ton of subcommands

**Review by roborourke — CHANGES_REQUESTED** [review]
I tried running `composer dev-tools codecept config` and got this:

```
[ErrorException]
Undefined offset 1
```

I also tried `composer dev-tools codecept gherkin:snippets acceptance` as examples and got:

```
Running "acceptance" test suite..
Error: Error establishing a database connection.
Starting headless "chrome" browser container..
a824b79b592b4048648d01fa4ec96eb10beaea5efb71e7bb2579c40de54d4e97
Running CodeCeption..
 All Gherkin steps are defined. Exiting... 
Removing headless browser container..
```

This is still a touch greedy in terms of how it runs and doesn't for more general use of codeception like generator commands etc. I think the custom handling should only apply for the `run` subsubcommand and the `bootstrap` subsubcommand as you have them now.

File: inc/command/class-command.php [hunk 10]
```diff
@@ -422,40 +430,63 @@ protected function codecept( InputInterface $input, OutputInterface $output ) {
...
+			$output->write(
+				sprintf( '<info>Detected %d suites, (%s)..</info>', count( $suites ), implode( ', ', $suites ) ),
+				true,
+				$output::VERBOSITY_NORMAL
+			);
+		}
 
-		// Ensure cache is clean.
-		$this->run_command( $input, $output, 'wp', [ 'cache', 'flush' ] );
+		$return = '';
```
Comment (roborourke) [inline inc/command/class-command.php]: Return value should be an integer exit code, so zero in this case if the command is considered successful so far.

File: inc/command/class-command.php (hunk 9 above)
Comment (shadyvb) [reply inc/command/class-command.php]: Nope, should be optional.

File: inc/command/class-command.php [hunk 11]
```diff
@@ -485,7 +487,12 @@ protected function codecept( InputInterface $input, OutputInterface $output ) {
 			}
 
 			$output->write( '<info>Running CodeCeption..</info>', true, $output::VERBOSITY_NORMAL );
-			$return = $this->run_command( $input, $output, 'vendor/bin/codecept', $options );
+			$return = $return ?: $this->run_command( $input, $output, 'vendor/bin/codecept', $options );
+
+			// Stop executing rest of the suites if one fails and `continue` argument is absent.
+			if ( $return && ! $input->getOption( 'continue' ) ) {
```
Comment (roborourke) [inline inc/command/class-command.php]: Missed this before but codeception accepts a comma separated list of suites, probably don't need the loop or this option, e.g. `composer dev-tools codecept run acceptance,functional,unit`.

File: inc/command/class-command.php (hunk 10 above)
Comment (shadyvb) [reply inc/command/class-command.php]: That inspired me to add a `-a` argument to continue running tests if one suite fails, and default to halting execution after a single failure.

File: inc/command/class-command.php (hunk 11 above)
Comment (shadyvb) [reply inc/command/class-command.php]: Codeception recommends _against_ running multiple suites in the same thread, hence most of the custom handling here, otherwise we would proxy the `run` command as is without any special handling. The way we did it here means end users can still use comma-separated suite names, but they'll work one by one, with each one checking for dependencies ( database import, browser container ) to minimize redundant actions.

Comment (roborourke) [reply inc/command/class-command.php]: Sorry you did say that. As long as we only hit this loop with the `run` command we should be good

**Review by roborourke — APPROVED** [review]
Nice work :)
